apache/incubator-mxnet
# [MXNET-107]Fused GRU implementation for CPU

## Description ##
In this PR, it aligns with #10104's registration way and creates a fused GRU operator for CPU.
@pengzhao-intel, @TaoLv , @Sherry-zhang 

## Feature changes ##
### New features ###
- Single layer/Multiple layer and unidirectional/bidirectional GRU, including both forward and backward computation.
- Support kAddTo/kNullOp check for both GRU and LSTM

### Unit-test changes ###
- Create new testcase in tests/python/unittests/test_operator.py.
- update testcase in example/rnn/bucketing/cudnn_rnn_bucketing.py
- Check consistency with original GRUCell implementation.

### Performance ###
We have tested performance of sym.RNN and rnn.GRUCell on our local Skylake-6148 with 2 Sockets and 40 cores. Use MKL as blas lib in this performance test.
Test input size is from DS2 default parameters(seq_length = 300, batch_size = 20, input_size = 800, hidden_size = 800).

Layer=1 bidirectional = False

| API | Inference time(fwd, sec)  |  Training time(fwd + bwd, sec)           |
| --------           | :-----:                              | :----:                                   |
| rnn.GRUCell  | 0.12                       |  0.277                            |
| this PR            | 0.048                      |  0.106                              |
| speedup         | 2.5x                              |  2.6x                                    | 


Layer=5 bidirectional = True

| API | Inference time(fwd, sec)  |  Training time(fwd + bwd, sec)           |
| --------           | :-----:                              | :----:                                   |
| rnn.GRUCell  | 1.275                       |  2.94                            |
| rnn.GRUCell (cuda)  | 1.047                       |  1.653                            |
| rnn.GRUCell (cudnn)  | 0.161                       |  0.442                           |
| this PR            | 0.551                      |  1.72                              |
| speedup -this PR/GRUCell         | 231.4%                              |  171%                                    | 
| speedup -this PR/GRUCell (cuda)        | 190%                              |  96%                                    | 
| speedup -this PR/GRUCell (cudnn)        | 29.2%                              |  25.7%                                   | 

### Convergency Curve ###
We have tested Convergency of FusedGRU on our CPU-Skylake-8180 with 2 Sockets and 56 cores and GPU-P100  by using example/rnn/bucketing/cudnn_rnn_bucketing.py
Test input size is layer = 3, batch_size = 32, num-embed = 800, num-hidden = 800, num-epochs 20
![fusedgru_cpu_gpu](https://user-images.githubusercontent.com/1567049/40525654-f24031f2-6013-11e8-8b5a-8c88b0e3c66c.png)


## Checklist ##
